### PR TITLE
Replace uuid dependency with internal library

### PR DESCRIPTION
**Pull request by stevvooe**
cc @tiborvass @dmcgowan 

Signed-off-by: Stephen J Day <stephen.day@docker.com>

### CODE REVIEW

File: uuid/uuid.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+// Generate creates a new, version 4 uuid.
+func Generate() (u UUID) {
+	for {
+		// This should never block but the read may fail. Because of this,
+		// we just try to read the random number generator until we get
+		// something. This is a very rare condition but may happen.
+
+		n, err := io.ReadFull(rand.Reader, u[:])
+		if err != nil {
+			logrus.Warnf("error generating version 4 uuid, retrying: %v", err)
```
Comment (tiborvass) [inline uuid/uuid.go]: Feels weird to have a logrus dependency in a uuid package. How about having a global variable that defaults to `func(msg string) { fmt.Fprintln(os.Stderr, msg) }` ? clients can then change it in their init() functions.

Comment (stevvooe) [inline uuid/uuid.go]: kk.

File: uuid/uuid.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	Size = Bits / 8
+
+	format = "%08x-%04x-%04x-%04x-%012x"
+)
+
+var (
+	// ErrUUIDInvalid indicates a parsed string is not a valid uuid.
+	ErrUUIDInvalid = fmt.Errorf("invalid uuid")
+
+	// Logger can be used to override the default logging destination. Such
```
Comment (tiborvass) [inline uuid/uuid.go]: `Loggerf`

**Comment by tiborvass** [comment]
LGTM

**Comment by stevvooe** [comment]
cc @NathanMcCauley

File: registry/storage/purgeuploads.go [hunk 3]
```diff
@@ -116,8 +116,8 @@ func getOutstandingUploads(ctx context.Context, driver storageDriver.StorageDriv
 func uUIDFromPath(path string) (string, bool) {
 	components := strings.Split(path, "/")
 	for i := len(components) - 1; i >= 0; i-- {
-		if uuid := uuid.Parse(components[i]); uuid != nil {
-			return uuid.String(), i == len(components)-1
+		if u, err := uuid.Parse(components[i]); err == nil {
```
Comment (dmcgowan) [inline registry/storage/purgeuploads.go]: Should this error be logged?

Comment (stevvooe) [inline registry/storage/purgeuploads.go]: No, this is just a validation.

**Comment by dmcgowan** [comment]
LGTM

**Comment by stevvooe** [comment]
Added retry backoff and restricted to EPERM. Will panic after 450ms of try to get entropy.

**Comment by NathanMcCauley** [comment]
LGTM

File: uuid/uuid_test.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			t.Fatalf("parsing round trip failed: %v != %v", parsed, u)
+		}
+	}
+
+	for _, c := range []string{
+		"bad",
+		"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",    // correct length, incorrect format
+		"  20cc7775-2671-43c7-8742-51d1cfa23258",  // leading space
+		"20cc7775-2671-43c7-8742-51d1cfa23258   ", // trailing space
+		"00000000-0000-0000-0000-x00000000000   ", // out of range character
```
Comment (thaJeztah) [inline uuid/uuid_test.go]: I don't think the trailing spaces should be here, otherwise you might be testing the wrong thing.

File: uuid/uuid.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+		// This should never block but the read may fail. Because of this,
+		// we just try to read the random number generator until we get
+		// something. This is a very rare condition but may happen.
+		b := time.Duration(retries) * backoff
+		time.Sleep(b)
+		totalBackoff += b
+
+		_, err := io.ReadFull(rand.Reader, u[:])
+		if err != nil {
+			if err == syscall.EPERM {
```
Comment (tiborvass) [inline uuid/uuid.go]: this syscall makes the "universal" part of UUID little less universal ;) maybe make the error handling platform specific? Or at least add a TODO for now.

Comment (tiborvass) [inline uuid/uuid.go]: apparently this does not prevent from compiling on windows, so i'm fine with it

**Comment by tiborvass** [comment]
LGTM

File: uuid/uuid.go (hunk 5 above)
Comment (noxiouz) [inline uuid/uuid.go]: I think, it fits better http://golang.org/pkg/os/#IsPermission
```go
os.IsPermission
```

**Pull request merged**